blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md

```diff
@@ -0,0 +1,247 @@
+1990. \[[Available online](https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node43.html#SECTION00700000000000000000)\]
+
+In chapter 3, Steele introduces the concepts of _scope_ and _extent_:
+
+> _Scope_ refers to the spatial or textual region of the program within which
+> references may occur. _Extent_ refers to the interval of time during which
+> references may occur.
+
+In addition, there are four interesting cases of scope and extent, with respect
+to Common Lisp:
+
+  - _Lexical scope_: a reference can only occur within certain textual regions
+    of the program, which are determined by the establishing construct, _e.g._,
+    the body of a variable definition.
+
+  - _Indefinite scope_: a reference can occur anywhere in the program.
+
+  - _Dynamic extent_: a reference can occur during the time between an entity's
+    creation and its explicit destruction, _e.g._, when a local variable is
+    created upon entering a function and destroyed when returning from that
+    function.
+
+  - _Indefinite extent_: an entity may exist as long as it is possible to be
+    referenced. (Note that this is the idea behind garbage collection: an
+    entity can be destroyed once references to it are impossible.)
+
+Steele points out that _dynamic scope_ is a misnomer, even though it is both a
+traditional and useful concept. It can be defined as _indefinite scope and
+dynamic extent_. In other words, references to a variable may occur anywhere in
+a program, as long as that variable has been initialized and has not yet been
+explicitly destroyed. Furthermore, a later initialization hides an earlier one.
+
+
+### Commentary
+
+I found this approach very informative, because it explicitly distinguishes
+between space (scope) and time (extent), which further implies a separation
+between compile time and run time. This explains my unease with the definition
+of "dynamic scope"---it is nominally about textual regions in a program, but
+also requires consideration of run-time behaviour. Dynamic scope is a misnomer!
+
+The above definitions are specifically for Common Lisp, but I believe we can
+learn from them and adapt them for other programming languages.
+
+
+## A brief and incomplete history of lexical scope
+
+During my research of different definitions of lexical scope, I began to wonder
+if there was an "original" definition of lexical scope. I did not find one, but
+I was able to trace some of the connections between Lisp, Scheme, and ALGOL 60.
+This history is certainly incomplete, but I hope it is somewhat useful and
+interesting.
+
+  * **1960**. John McCarthy publishes the original paper on Lisp.[^2] In
+    _History of Lisp_,[^3] McCarthy writes that he borrowed the λ-notation from
+    Alonzo Church's lambda calculus, but none of the other ideas. He also
+    recounts an incident where a programmer desired lexical scoping, but Lisp
+    used dynamic scoping. This was later solved by the "FUNARG device."
```

Comment:
Worth noting that on the same page, McCarthy says he considered dynamic scope a bug?

Reply:
Excellent suggestion, thanks!

```diff
     recounts an incident where a programmer desired lexical scoping, but Lisp
-    used dynamic scoping. This was later solved by the "FUNARG device."
+    used dynamic scoping. McCarthy considered this to be a bug, which Steve
+    Russell later fixed by developing the "FUNARG device."
```